File: Snap-O/ADB/DeviceTracker.swift
import Foundation

private let log = SnapOLog.tracker

final class DeviceTracker: @unchecked Sendable {
  private let adbService: ADBService

  private var trackTask: Task<Void, Never>?
  private var continuations: [UUID: AsyncStream<[Device]>.Continuation] = [:]
  private let lock = NSLock()
  private var _latestDevices: [Device] = []
  var latestDevices: [Device] {
    lock.lock()
    defer { lock.unlock() }
    return _latestDevices
  }

  private var infoCache: [String: DeviceInfo] = [:]
  private var hasSeenFirstMessage: Bool = false

  init(adbService: ADBService) {
    self.adbService = adbService
  }

  // MARK: - Public API

  func deviceStream() -> AsyncStream<[Device]> {
    let id = UUID()
    return AsyncStream { continuation in
      lock.lock()
      continuations[id] = continuation
      lock.unlock()
      if self.hasSeenFirstMessage {
        continuation.yield(self.latestDevices)
      }

      continuation.onTermination = { [weak self] _ in
        self?.removeContinuation(id)
      }
    }
  }

  // MARK: - Tracking

  func startTracking() {
    trackTask?.cancel()
    trackTask = Task { [weak self] in
      await self?.trackLoop()
    }
  }

  private func removeContinuation(_ id: UUID) {
    continuations.removeValue(forKey: id)
  }

  private func broadcast(_ devices: [Device]) {
    lock.lock()
    _latestDevices = devices
    hasSeenFirstMessage = true
    let snapshot = Array(continuations.values)
    lock.unlock()
    for continuation in snapshot {
      continuation.yield(devices)
    }
  }

  private func trackLoop() async {
    @inline(__always)
    func pause() async {
      try? await Task.sleep(for: .milliseconds(300))
    }

    while !Task.isCancelled {
      let exec = await adbService.exec()
      guard let (handle, stream) = try? await exec.trackDevices() else {
        if hasSeenFirstMessage { broadcast([]) }
        await pause()
        continue
      }

      defer { handle.cancel() }

      do {
        for try await payload in stream {
          if Task.isCancelled { break }
          let devices = await parseDevices(from: payload, exec: exec)
          broadcast(devices)
        }
        await pause()
      } catch is CancellationError {
        break
      } catch {
        if hasSeenFirstMessage { broadcast([]) }
        await pause()
      }
    }
  }

  // MARK: - Device parsing

  private func parseDevices(from payload: String, exec: ADBExec) async -> [Device] {
    let parsed = payload
      .split(separator: "\n", omittingEmptySubsequences: true)
      .compactMap(parseDeviceRow)

    return await withTaskGroup(of: Device?.self) { group in
      for (id, fields) in parsed {
        group.addTask {
          let info = await self.deviceInfo(for: id, fallbackModel: fields["model"], exec: exec)
          return Device(
            id: id,
            model: info.model,
            androidVersion: info.version,
            vendorModel: info.vendorModel,
            manufacturer: info.manufacturer,
            avdName: info.avdName
          )
        }
      }
      var out: [Device] = []
      for await device in group {
        if let device { out.append(device) }
      }
      return out
    }
  }

  /// Parses a single `adb devices -l` row like:
  ///   `<serial> device product:foo model:Pixel_7 device:panther transport_id:3`
  /// Returns nil for headers, empties, or unwanted states (offline/unauthorized).
  private func parseDeviceRow(_ line: Substring) -> (id: String, fields: [String: String])? {
    let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else { return nil }

    let parts = trimmed.split(whereSeparator: \.isWhitespace)
    guard let first = parts.first else { return nil }
    let id = String(first)

    if parts.count >= 2 {
      let state = parts[1].lowercased()
      if state.contains("offline") || state.contains("unauthorized") || state.contains("recovery") {
        return nil
      }
    }

    // Parse key:value pairs into a dictionary.
    var fields: [String: String] = [:]
    fields.reserveCapacity(6)
    for part in parts.dropFirst() {
      if let idx = part.firstIndex(of: ":") {
        let key = String(part[..<idx])
        let value = String(part[part.index(after: idx)...])
        fields[key] = value
      }
    }

    return (id, fields)
  }

  private func deviceInfo(for id: String, fallbackModel: String?, exec: ADBExec) async -> DeviceInfo {
    if let cached = infoCache[id] {
      return cached
    }

    // Single getprop dump and extract the properties we care about
    let props = await (try? exec.getProperties(deviceID: id, prefix: "ro.")) ?? [:]

    let model = fallbackModel
      ?? cleanProp("ro.product.model", in: props)
      ?? "Unknown Model"
    let version = cleanProp("ro.build.version.release", in: props) ?? "Unknown API"
    let vendorModel = cleanProp("ro.product.vendor.model", in: props)
    let manufacturer = cleanProp("ro.product.vendor.manufacturer", in: props)
      ?? cleanProp("ro.product.manufacturer", in: props)
    let avdName = cleanProp("ro.boot.qemu.avd_name", in: props)
      .map { $0.replacingOccurrences(of: "_", with: " ") }

    let info = DeviceInfo(
      model: model,
      version: version,
      vendorModel: vendorModel,
      manufacturer: manufacturer,
      avdName: avdName
    )
    infoCache[id] = info
    return info
  }

  // MARK: - Helpers

  private struct DeviceInfo {
    let model: String
    let version: String
    let vendorModel: String?
    let manufacturer: String?
    let avdName: String?
  }

  // MARK: - Property helpers

  private func cleanProp(_ key: String, in props: [String: String]) -> String? {
    guard let raw = props[key]?.trimmingCharacters(in: .whitespacesAndNewlines) else { return nil }
    return raw.isEmpty ? nil : raw
  }
}

File: snapo-desktop-compose/src/main/kotlin/com/openai/snapo/desktop/adb/DeviceTracker.kt
package com.openai.snapo.desktop.adb

import com.openai.snapo.desktop.di.AppScope
import dev.zacsweers.metro.Inject
import dev.zacsweers.metro.SingleIn
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.withContext

@SingleIn(AppScope::class)
@Inject
class DeviceTracker(
    private val adb: AdbExec,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default),
) {
    private val infoCache = HashMap<String, DeviceInfo>()

    val devices: StateFlow<List<Device>> = flow {
        suspend fun pause() = delay(300)

        while (true) {
            try {
                adb.trackDevices().collect { payload ->
                    emit(parseDevices(payload))
                }
                pause()
            } catch (t: CancellationException) {
                throw t
            } catch (_: Throwable) {
                emit(emptyList())
                pause()
            }
        }
    }.stateIn(
        scope,
        SharingStarted.WhileSubscribed(stopTimeoutMillis = 1_000),
        emptyList(),
    )

    val latestDevices: List<Device>
        get() = devices.value

    private suspend fun parseDevices(payload: String): List<Device> {
        val parsedRows = payload
            .lineSequence()
            .mapNotNull { parseDeviceRow(it) }
            .toList()

        // Fetch ro.* props once per device ID and cache.
        return parsedRows.map { (id, fields) ->
            val info = deviceInfoFor(id, fallbackModel = fields["model"])
            Device(
                id = id,
                model = info.model,
                androidVersion = info.version,
                vendorModel = info.vendorModel,
                manufacturer = info.manufacturer,
                avdName = info.avdName,
            )
        }
    }

    private fun parseDeviceRow(line: String): Pair<String, Map<String, String>>? {
        val trimmed = line.trim()
        if (trimmed.isEmpty()) return null

        val parts = trimmed.split(Regex("\\s+"))
        val id = parts.firstOrNull()?.trim().orEmpty()
        if (id.isEmpty()) return null

        if (parts.size >= 2) {
            val state = parts[1].lowercase()
            val invalidStates = listOf("offline", "unauthorized", "recovery", "authorizing")
            if (invalidStates.any(state::contains)) {
                return null
            }
        }

        val fields = LinkedHashMap<String, String>(6)
        for (part in parts.drop(1)) {
            val idx = part.indexOf(':')
            if (idx <= 0 || idx == part.lastIndex) continue
            val key = part.substring(0, idx)
            val value = part.substring(idx + 1)
            fields[key] = value
        }
        return id to fields
    }

    private suspend fun deviceInfoFor(id: String, fallbackModel: String?): DeviceInfo {
        infoCache[id]?.let { return it }

        val props = try {
            withContext(Dispatchers.IO) { adb.getProperties(deviceId = id, prefix = "ro.") }
        } catch (_: Throwable) {
            emptyMap()
        }

        fun cleanProp(key: String): String? = props[key]?.trim()?.takeIf { it.isNotEmpty() }

        val model = fallbackModel
            ?.replace('_', ' ')
            ?.takeIf { it.isNotBlank() }
            ?: cleanProp("ro.product.model")
            ?: "Unknown Model"

        val version = cleanProp("ro.build.version.release") ?: "Unknown API"

        val vendorModel = cleanProp("ro.product.vendor.model")
        val manufacturer = cleanProp("ro.product.vendor.manufacturer")
            ?: cleanProp("ro.product.manufacturer")

        val avdName = cleanProp("ro.boot.qemu.avd_name")
            ?.replace('_', ' ')

        val info = DeviceInfo(
            model = model,
            version = version,
            vendorModel = vendorModel,
            manufacturer = manufacturer,
            avdName = avdName,
        )
        infoCache[id] = info
        return info
    }

    private data class DeviceInfo(
        val model: String,
        val version: String,
        val vendorModel: String?,
        val manufacturer: String?,
        val avdName: String?,
    )
}
